core/modules/admin_bar/admin_bar.inc

```diff
@@ -109,13 +107,50 @@ function admin_bar_output($complete = FALSE) {
 
   // Store the new hash for this user.
   if (!$complete) {
-    admin_bar_cache_set($cid, md5($content));
+    admin_bar_cache_set($cid, md5(json_encode($content)));
   }
+  // Don't cache the alerts
+  $content['extra']['extra'] += admin_bar_links_alerts();
 
+  $content = backdrop_render($content);
   return $content;
 }
 
 /**
+ * Build an alert icon with dropdown links.
+ */
+function admin_bar_links_alerts() {
+  // Get all module provided alerts.
+  $alerts = module_invoke_all('admin_bar_alerts');
+  
+  // Find out if any modules provided an 'error' or 'warning' alert.
+  $levels = array_column($alerts, 'severity');
```

Comment:
the `array_column` function is only available in PHP 5.5 and greater. We still support 5.3 so we'll need to use a different function here.